include/maidsafe/crux/detail/multiplexer.hpp

```diff
@@ -369,39 +387,35 @@ inline void multiplexer::do_start_receive()
 {
     auto self(shared_from_this());
 
-    // We need to read with at least one zero sized buffer to
-    // get the remote_endpoint information.
-    next_layer().async_receive_from
-        (boost::asio::buffer(static_cast<char*>(nullptr), 0),
-         next_remote_endpoint,
-         std::remove_reference<decltype(next_layer())>::type::message_peek,
+    next_layer().async_receive
+        (boost::asio::null_buffers(),
          [self]
          (boost::system::error_code error, std::size_t /*size*/) mutable
          {
-            // The size parameter is useless here because what we get
-            // is min(buffer_size, datagram_size) and our buffer size is 0.
-            self->process_peek(error, self->next_remote_endpoint);
+            self->process_new_message(error);
```

Comment:
I wasn't aware about the null_buffers guarantee, very good to know, thanks. I am not sure about the second argument, we could have the header that goes to async_peek(header) as a member of the multiplexer because we always do only one udp::socket::async_receive operation at a time, thus we would avoid the header data allocation. The empty message could still be just two bytes (the tag), we would get the size information in the handler of the async_peek(header) operation.